### PR TITLE
containerd integration: Implement ImageDelete for containerd

### PR DESCRIPTION
**Pull request by thaJeztah**
- ~depends on https://github.com/moby/moby/pull/43818~
- upstreaming https://github.com/rumpl/moby/pull/11

**- What I did**

Implemented image deletion for containerd

### CODE REVIEW

File: daemon/containerd/image_delete.go [hunk 1]
```diff
@@ -35,6 +41,24 @@ import "github.com/docker/docker/api/types"
 // If prune is true, ancestor images will each attempt to be deleted quietly,
 // meaning any delete conflicts will cause the image to not be deleted and the
 // conflict will not be reported.
-func (i *ImageService) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
-	panic("not implemented")
+func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
+	// FIXME(thaJeztah): implement ImageDelete "force"
+	// FIXME(thaJeztah): implement ImageDelete "prune"
```
Comment (thaJeztah) [inline daemon/containerd/image_delete.go:46]: Let me create tracking issues for these

Comment (thaJeztah) [reply daemon/containerd/image_delete.go]: Created:

- https://github.com/moby/moby/issues/43849
- https://github.com/moby/moby/issues/43850

File: daemon/containerd/image_delete.go [hunk 2]
```diff
@@ -35,6 +41,24 @@ import "github.com/docker/docker/api/types"
 // If prune is true, ancestor images will each attempt to be deleted quietly,
 // meaning any delete conflicts will cause the image to not be deleted and the
 // conflict will not be reported.
-func (i *ImageService) ImageDelete(imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
-	panic("not implemented")
+func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]types.ImageDeleteResponseItem, error) {
+	// TODO(thaJeztah): implement ImageDelete "force" options; see https://github.com/moby/moby/issues/43850
+	// TODO(thaJeztah): implement ImageDelete "prune" options; see https://github.com/moby/moby/issues/43849
```
Comment (thaJeztah) [inline daemon/containerd/image_delete.go:46]: Per discussion in the maintainers meeting; we need to add more tracking issues for; partial IDs etc. (some of that logic is implemented in `GetImage()` as part of https://github.com/moby/moby/pull/43818)

**Comment by thaJeztah** [comment]
@corhere @tianon @rumpl PTAL

**Review by rumpl — APPROVED** [review]

File: daemon/containerd/image_delete.go [hunk 3]
```diff
@@ -35,6 +41,24 @@ import "github.com/docker/docker/api/types"
...
+
+	records := []types.ImageDeleteResponseItem{}
```
Comment (corhere) [inline daemon/containerd/image_delete.go:48]: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

File: daemon/containerd/image_delete.go [hunk 4]
```diff
@@ -35,6 +41,24 @@ import "github.com/docker/docker/api/types"
...
+
+	parsedRef, err := reference.ParseNormalizedNamed(imageRef)
+	if err != nil {
+		return nil, err
+	}
+	ref := reference.TagNameOnly(parsedRef)
+
+	if err := i.client.ImageService().Delete(ctx, ref.String(), images.SynchronousDelete()); err != nil {
+		return []types.ImageDeleteResponseItem{}, err
```
Comment (corhere) [inline daemon/containerd/image_delete.go:57]: How come an empty slice is returned here but the error-return for `ParseNotmalizedNamed` (line 52) returns `nil`?

File: daemon/images/image_prune.go [hunk 5]
```diff
@@ -119,7 +119,7 @@ deleteImagesLoop:
 
 			if shouldDelete {
 				for _, ref := range refs {
-					imgDel, err := i.ImageDelete(ref.String(), false, true)
+					imgDel, err := i.ImageDelete(ctx, ref.String(), false, true)
 					if imageDeleteFailed(ref.String(), err) {
```
Comment (corhere) [inline daemon/images/image_prune.go:123]: Looks like `imageDeleteFailed()` needs to be updated so it doesn't log a spurious warning when the context is canceled.

**Review by corhere — CHANGES_REQUESTED** [review]

File: daemon/containerd/image_delete.go (hunk 3 above)
Comment (thaJeztah) [reply daemon/containerd/image_delete.go:48]: Hm, yeah, I don't _always_ agree with that rule (sometimes it's really more convenient to initialise it if later on we need to `append()`, and `make(...)` is a bit cumbersome (or sometimes overdoing it).

That said; it looks like (at least currently) it'll only ever create a single instance, so I don't think we need this variable at all; we can just return `[]types.ImageDeleteResponseItem{}{d}` at the end of the function.

File: daemon/containerd/image_delete.go (hunk 4 above)
Comment (thaJeztah) [reply daemon/containerd/image_delete.go:57]: Good catch; at a glance, I think can be just `nil` as well, unless there's some expectation somewhere (`nil` vs `nil type`)? Overall the returned value shouldn't be used if there's an error.

Let me have a look at that.

File: daemon/containerd/image_delete.go (hunk 3 above)
Comment (corhere) [reply daemon/containerd/image_delete.go:48]: > sometimes it's really more convenient to initialise it if later on we need to `append()`

Hmm? `append()` works on `nil` slices, same as non-`nil` but empty slices. You might be thinking of maps, where assigning to a `nil` map panics.

Comment (thaJeztah) [reply daemon/containerd/image_delete.go]: Yes, I just edited, indeed had maps in mind 🙈

File: daemon/images/image_prune.go [hunk 6]
```diff
@@ -163,7 +163,7 @@ func imageDeleteFailed(ref string, err error) bool {
 	switch {
 	case err == nil:
 		return false
-	case errdefs.IsConflict(err):
+	case errdefs.IsConflict(err), errors.Is(err, context.Canceled):
```
Comment (corhere) [inline daemon/images/image_prune.go:166]:
```suggestion
	case errdefs.IsConflict(err), errors.Is(err, context.Canceled), errors.Is(err, context.DeadlineExceeded):
```
Annoyingly, you need to enumerate all the possible context errors.

**Review by corhere — APPROVED** [review]

File: daemon/images/image_prune.go (hunk 6 above)
Comment (thaJeztah) [reply daemon/images/image_prune.go:166]: LOL; I actually had the DeadlineExceeded, and started to think if that one would be one to actually _log_, so removed it; let me add it back 😂

Comment (thaJeztah) [reply daemon/images/image_prune.go]: And it's back (it was _literally_ `CMD+Z` - last edit I made 😂 )

**Comment by thaJeztah** [comment]
@AkihiroSuda @tonistiigi PTAL

**Review by cpuguy83 — APPROVED** [review]

**Pull request merged**